File: apps/android/app/src/main/java/com/pawwalk/android/ui/screens/HomeViewModel.kt
package com.pawwalk.android.ui.screens

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.pawwalk.android.data.Booking
import com.pawwalk.android.data.BookingRepository
import com.pawwalk.android.data.OwnerStats
import com.pawwalk.android.data.Network
import com.pawwalk.android.data.Pet
import com.pawwalk.android.data.PetRepository
import com.pawwalk.android.data.Walker
import com.pawwalk.android.data.WalkerRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Duration
import java.time.Instant

/**
 * Backs HomeScreen with the signed-in user's real bookings (mirrors the iOS
 * HomeViewModel — see docs/FUNCTIONAL-REVIEW.md N3). Derives the "next walk" and
 * a this-week count; distance/streak/recent walks come from GET /bookings/stats.
 */
class HomeViewModel : ViewModel() {
    data class UpcomingWalk(val booking: Booking, val walker: Walker?)
    data class HomeState(
        val upcoming: UpcomingWalk? = null,
        val weekCount: Int = 0,
        val pets: List<Pet> = emptyList(),
        val stats: OwnerStats? = null,
    )

    private val _state = MutableStateFlow(HomeState())
    val state: StateFlow<HomeState> = _state.asStateFlow()

    init { load() }

    fun load() {
        viewModelScope.launch {
            val bookings = try { BookingRepository.fetchBookings() } catch (e: Exception) { emptyList() }
            val byId = runCatching { WalkerRepository.fetchWalkers() }
                .getOrDefault(emptyList()).associateBy { it.id }
            val pets = runCatching { PetRepository.list() }.getOrDefault(emptyList())
            val stats = runCatching { Network.api.ownerStats() }.getOrNull()
            val now = Instant.now()
            val future = bookings
                .filter { it.isActive }
                .mapNotNull { b -> runCatching { Instant.parse(b.startTime) }.getOrNull()?.let { b to it } }
                .filter { (_, start) -> !start.isBefore(now) }
                .sortedBy { (_, start) -> start }
            val upcoming = future.firstOrNull()?.let { (b, _) -> UpcomingWalk(b, byId[b.walkerId]) }
            val weekAhead = now.plus(Duration.ofDays(7))
            val weekCount = future.count { (_, start) -> !start.isAfter(weekAhead) }
            _state.value = HomeState(upcoming, weekCount, pets, stats)
        }
    }
}

File: apps/ios/PawWalk/Features/Home/HomeViewModel.swift
import Foundation
import Observation

/// Backs `HomeView` with the signed-in user's real data (see
/// docs/FUNCTIONAL-REVIEW.md N3). Loads bookings + walkers + pets and the
/// server-computed stats (tracked distance, streak, recent walks), and derives
/// the "next walk" and a this-week count.
@MainActor
@Observable
final class HomeViewModel {
    private(set) var upcoming: (booking: Booking, walker: Walker?)?
    private(set) var weekCount = 0
    private(set) var pets: [Pet] = []
    private(set) var stats: OwnerStats?
    private(set) var loaded = false

    func load() async {
        async let bookingsCall = APIClient.shared.bookings()
        async let walkersCall = APIClient.shared.walkers()
        async let petsCall = APIClient.shared.pets()
        async let statsCall = APIClient.shared.ownerStats()
        let bookings = (try? await bookingsCall) ?? []
        let walkers = (try? await walkersCall) ?? []
        pets = (try? await petsCall) ?? []
        stats = try? await statsCall
        let byID = Dictionary(uniqueKeysWithValues: walkers.map { ($0.id, $0) })

        let now = Date()
        let future = bookings
            .filter { $0.status != .cancelled && $0.startTime >= now }
            .sorted { $0.startTime < $1.startTime }

        upcoming = future.first.map { ($0, byID[$0.walkerID]) }
        let weekAhead = now.addingTimeInterval(7 * 24 * 3600)
        weekCount = future.filter { $0.startTime <= weekAhead }.count
        loaded = true
    }
}
